components/homme/src/share/cxx/SphereOperators.hpp

```diff
@@ -1176,6 +1176,279 @@ class SphereOperators
+    const int ixy = tr / SPHERE_BLOCK_LEV;
+    x = ixy / NP;
+    y = ixy % NP;
+    const int dz = tr % SPHERE_BLOCK_LEV;
+    z = dz + iw * SPHERE_BLOCK_LEV;
+
+    metdet = g.metdet(e,x,y);
+    for (int i = 0; i < 2; i++) for (int j = 0; j < 2; j++) dinv[i][j] = g.dinv(e,i,j,x,y);
+    for (int j = 0; j < NP; j++) {
+      dvvx[j] = g.dvv(x,j);
+      dvvy[j] = g.dvv(y,j);
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION void barrier() const
+  {
+    t.team_barrier();
+  }
+
+  KOKKOS_INLINE_FUNCTION Real div(const SphereBlockScratch &t0, const SphereBlockScratch &t1)
+  {
+    // Separately compute du and dv to match Fortran bfb
+    Real du = 0;
+    Real dv = 0;
+    for (int j = 0; j < NP; j++) {
+      du += dvvy[j] * t0.sv(x,j);
+      dv += dvvx[j] * t1.sv(j,y);
+    }
+    if (rrdmd == 0) rrdmd = (1.0 / metdet) * scale_factor_inv;
+    return (du + dv) * rrdmd;
+  }
+
+  KOKKOS_INLINE_FUNCTION void divInit(SphereBlockScratch &t0, SphereBlockScratch &t1, const Real v0, const Real v1) const
+  {
+    t0.sv(x,y) = (dinv[0][0] * v0 + dinv[1][0] * v1) * metdet;
+    t1.sv(x,y) = (dinv[0][1] * v0 + dinv[1][1] * v1) * metdet;
+  }
+
+  KOKKOS_INLINE_FUNCTION void grad(Real &g0, Real &g1, const SphereBlockScratch &t) const
+  {
+    Real s0 = 0;
+    Real s1 = 0;
+    for (int j = 0; j < NP; j++) {
+      s0 += dvvy[j] * t.sv(x,j);
+      s1 += dvvx[j] * t.sv(j,y);
+    }
+    s0 *= scale_factor_inv;
+    s1 *= scale_factor_inv;
+    g0 = dinv[0][0] * s0 + dinv[0][1] * s1;
+    g1 = dinv[1][0] * s0 + dinv[1][1] * s1;
+  }
+
+  KOKKOS_INLINE_FUNCTION void gradInit(SphereBlockScratch &t0, const Real v0)
+  {
+    t0.sv(x,y) = v0;
+  }
+
+  KOKKOS_INLINE_FUNCTION bool skip() const { return (z >= NUM_PHYSICAL_LEV); }
+
+  KOKKOS_INLINE_FUNCTION Real vort(const SphereBlockScratch &t0, const SphereBlockScratch &t1)
+  {
+    Real du = 0;
+    Real dv = 0;
+    for (int j = 0; j < NP; j++) {
+      du += dvvx[j] * t0.sv(j,y);
+      dv += dvvy[j] * t1.sv(x,j);
+    }
+    if (rrdmd == 0) rrdmd = (1.0 / metdet) * scale_factor_inv;
+    return (dv - du) * rrdmd;
+  }
+
+  KOKKOS_INLINE_FUNCTION void vortInit(SphereBlockScratch &t0, SphereBlockScratch &t1, const Real v0, const Real v1) const
+  {
+    t0.sv(x,y) = g.d(e,0,0,x,y) * v0 + g.d(e,0,1,x,y) * v1;
+    t1.sv(x,y) = g.d(e,1,0,x,y) * v0 + g.d(e,1,1,x,y) * v1;
+  }
+
+  static TeamPolicy policy(const int num_elems, const int num_scratch)
+  {
+    return TeamPolicy(num_elems * SPHERE_BLOCKS_PER_COL, SPHERE_BLOCK).
+      set_scratch_size(0, Kokkos::PerTeam(num_scratch * SphereBlockScratchView::shmem_size()));
+  }
+
+};
+
+KOKKOS_INLINE_FUNCTION SphereBlockScratch::SphereBlockScratch(const SphereBlockOps &b):
+  v(b.t.team_scratch(0)),
+  sv(Kokkos::subview(v, b.z % SPHERE_BLOCK_LEV, Kokkos::ALL, Kokkos::ALL))
+{}
+
+struct SphereCol {
+  const Team &t;
+  int e,x,y,z;
+
+  KOKKOS_INLINE_FUNCTION SphereCol(const Team &team):
+    t(team)
+  {
+    const int lr = t.league_rank();
```

Comment:
Quick question: this seems to assume that the team policy has league size `num_elems*NP*NP`, right? If so, see my other comment in CaarFunctorImpl.cpp...